gk/main.c

```diff
@@ -552,6 +552,72 @@ gk_setup_rss(struct gk_config *gk_conf)
 	return ret;
 }
 
+static void
+gk_arp_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	unsigned int i;
+	struct gk_config *gk_conf = arg;
+	struct gk_lpm *ltbl = gk_conf->lpm_tbl;
+
+	for (i = 0; i < ltbl->num_tunnels; i++) {
+		if (ltbl->tunnels[i].flow.proto == ETHER_TYPE_IPv4
+				&& memcmp(&ltbl->tunnels[i].flow.f.v6.dst,
+				map->ip_be, sizeof(struct in_addr)) == 0) {
+			ether_addr_copy(&map->ha,
+				&ltbl->tunnels[i].nexthop_mac);
+			break;
+		}
+	}
+
+	*pcall_again = true;
+}
+
+static void
+gk_nd_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	unsigned int i;
+	struct gk_config *gk_conf = arg;
+	struct gk_lpm *ltbl = gk_conf->lpm_tbl;
+
+	for (i = 0; i < ltbl->num_tunnels; i++) {
+		if (ltbl->tunnels[i].flow.proto == ETHER_TYPE_IPv6
+				&& memcmp(ltbl->tunnels[i].flow.f.v6.dst,
+				map->ip_be, sizeof(struct in6_addr)) == 0) {
+			ether_addr_copy(&map->ha,
+				&ltbl->tunnels[i].nexthop_mac);
+			break;
+		}
+	}
+
+	*pcall_again = true;
+}
+
+static int
+gk_hold_arp_and_nd(struct gk_config *gk_conf, unsigned int lcore_id)
+{
+	unsigned int i;
+	int ret;
+	struct gk_lpm *ltbl = gk_conf->lpm_tbl;
+
+	for (i = 0; i < ltbl->num_tunnels; i++) {
+		if (ltbl->tunnels[i].flow.proto == ETHER_TYPE_IPv4) {
+			ret = hold_arp(gk_arp_req_cb, (void *)gk_conf, (struct in_addr *)
+				&ltbl->tunnels[i].flow.f.v4.dst, lcore_id);
```

Comment:
There is a structural bug at the call `hold_arp()` above and `hold_nd()` below that requires changes in both patches of this pull request.

The `struct ipip_tunnel_info`, which's the type of `ltbl->tunnels[i]`, holds the address of the Gatekeeper server at the source address, holds the address of the Grantor server at the destination address, holds the MAC address of the back interface at the source MAC address, AND supposedly holds the MAC address of the Grantor server.

Notice that for virtually all deployments, the Grantor server will be in a network different from the network of the Gatekeeper server, so the Gatekeeper server will never resolve the IP address of the Grantor server to a local MAC address.

The only way to solve this problem is for the Gatekeeper to resolve the Grantor server through the LPM table to find the gateway to use to send the packet.